### PR TITLE
Support extra subsidy for cow token owners

### PR DESCRIPTION
**Pull request by vkgnosis**
This is the main work for https://github.com/gnosis/gp-v2-services/issues/1647 . We support configuring the service with an erc20 token and extra subsidy for users that have a minimum balance in this token.

The model is a simple binary yes if the threshold is exceed for multiplying an extra factor. We could extend it with other functions in the future for example to scale the subsidy with the amount held.

I have not yet looked at the actual cow token contract so might need some adjustments with that regard in another PR.

In the default configuration no fee subsidy is given.

### Test Plan

new unit test and new ignored test

### CODE REVIEW

File: crates/orderbook/src/fee.rs [hunk 1]
```diff
@@ -322,7 +351,21 @@ impl MinFeeCalculating for MinFeeCalculator {
...
+        let cow_owner = async {
+            match user {
+                Some(user) => self.cow_subsidy.qualifies_for_subsidy(user).await,
+                None => Ok(false),
+            }
+        };
+        let past_fee = self
+            .measurements
+            .find_measurement_including_larger_amount(fee_data, (self.now)());
+        // TODO: make this function return a proper error distinguishing errors from fee too low;
```
Comment (vkgnosis) [inline crates/orderbook/src/fee.rs:365]: Want to do this in another PR

File: crates/orderbook/src/fee.rs [hunk 2]
```diff
@@ -150,6 +159,9 @@ impl From<u32> for FeeParameters {
 #[cfg_attr(test, mockall::automock)]
 #[async_trait::async_trait]
 pub trait MinFeeCalculating: Send + Sync {
+    // TODO: User should be mandatory. It is not because some old routes like /api/v1/fee do not
+    // know the user. When we remove those apis we can remove the Optional.
+
```
Comment (nlordell) [inline crates/orderbook/src/fee.rs:164]: Instead, can't the `api/v1/fee` just use `H160::default()`? Note that `address(0)` is guaranteed to never have a balance:

https://github.com/gnosis/cow-token/blob/6cc9c8ba02116869b319a705f933bcad99633529/src/contracts/vendored/mixins/ERC20.sol#L272-L273

**Review by nlordell — COMMENTED** [review]
LGTM.

I had some concerns about this making the fee computation time longer, but since we `join` with the price estimates, it will very likely not be the bottleneck.

I also think that `user` should not be optional. We should be able to safely use `H160::default()` in the legacy route because the contract prevents transferring tokens to that address anyway (same for the COW token).

File: crates/orderbook/src/fee.rs (hunk 2 above)
Comment (vkgnosis) [reply crates/orderbook/src/fee.rs:164]: I was testing with weth where the 0 address actually has a lot of balance so I thought optional was better for now. Since the cow token does not allow this agree we can use it instead of None.

File: crates/model/src/order.rs [hunk 3]
```diff
@@ -39,13 +39,19 @@ pub struct Order {
...
 impl Default for Order {
     fn default() -> Self {
+        let domain = &DomainSeparator::default();
+        let order = OrderCreation::default();
+        let owner = order
+            .signature
+            .validate(domain, &order.hash_struct())
+            .unwrap();
         Self::from_order_creation(
             OrderCreation::default(),
```
Comment (sunce86) [inline crates/model/src/order.rs:49]: nit: Can we reuse the already created `order` here?

File: crates/orderbook/src/fee.rs (hunk 2 above)
Comment (vkgnosis) [reply crates/orderbook/src/fee.rs]: Done

**Review by sunce86 — APPROVED** [review]
The current implementation LGTM, however I was under impression that the idea for cow token subsidy is not finalized yet (still under discussion in the forum) and could make this impl obsolete.

File: crates/orderbook/src/cow_subsidy.rs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+use contracts::ERC20;
+use primitive_types::{H160, U256};
+
+const CACHE_SIZE: usize = 10000;
+const CACHE_LIFESPAN: Duration = Duration::from_secs(60 * 60);
+
+/// Check whether a user qualifies for an extra fee subsidy because they own enough cow token.
+#[async_trait::async_trait]
+pub trait CowSubsidy: Send + Sync + 'static {
+    async fn qualifies_for_subsidy(&self, user: H160) -> Result<bool>;
```
Comment (nlordell) [inline crates/orderbook/src/cow_subsidy.rs:14]: Does it make sense for this component to be a simple `CowBalance` component to keep all subsidy logic in `fee.rs`?

The main reason I was suggesting this is that I'm 90% sure that we will want this to be tiered so we will need the balance amount and not a binary "has balance or not". Additionally, it feels a bit weird to me that the subsidy threshold is configured here and the subsidy amount is configured by the `MinFeeCalculating` implementation (i.e. the code could be a bit more colocated).

**Review by nlordell — COMMENTED** [review]
In general looks good. Just some comments about how the logic is split.

File: crates/orderbook/src/cow_subsidy.rs (hunk 4 above)
Comment (vkgnosis) [reply crates/orderbook/src/cow_subsidy.rs:14]: Alternatively we can store the threshold and factor both in this component and have `fn cow_subsidy(&self, user: H160) -> Result<f64>`. One difference is that having all the cow subsidy related code is in one place then so that the balance fetching is not separate from the fee calculation. It could be good that this keeps the fee component simpler. This would be reflected in the cache too as at the moment the cache can store `bool` but in your example it would store `U256`.
I'm not sure what is better.

Comment (nlordell) [reply crates/orderbook/src/cow_subsidy.rs:14]: Hmm, good point. I guess it does make sense to keep the fee component simpler. Then I would suggest `fn cow_subsidy(&self, user: H160) -> Result<f64>` so that COW related subsidy code is in one place.

File: crates/orderbook/src/cow_subsidy.rs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+const CACHE_SIZE: usize = 10000;
+const CACHE_LIFESPAN: Duration = Duration::from_secs(60 * 60);
+
+/// Check whether a user qualifies for an extra fee subsidy because they own enough cow token.
+#[async_trait::async_trait]
+pub trait CowSubsidy: Send + Sync + 'static {
+    async fn cow_subsidy_factor(&self, user: H160) -> Result<f64>;
+}
+
+pub struct NoopCowSubsidy(pub f64);
```
Comment (nlordell) [inline crates/orderbook/src/cow_subsidy.rs:17]: naming-nit: 
```suggestion
pub struct FixedCowSubsidy(pub f64);
```

**Review by nlordell — APPROVED** [review]
LGTM
